R/Spectra-functions.R

```diff
@@ -623,4 +588,91 @@ joinSpectraData <- function(x, y,
 #' @rdname Spectra
 processingLog <- function(x) {
     x@processing
-}
\ No newline at end of file
+}
+
+#' @export
+#'
+#' @rdname Spectra
+estimatePrecursorIntensity <- function(x, ppm = 10, tolerance = 0,
+                                       method = c("previous", "interpolation"),
+                                       msLevel. = 2L, f = dataOrigin(x),
+                                       BPPARAM = bpparam()) {
+    if (is.factor(f))
+        f <- as.character(f)
+    f <- factor(f, levels = unique(f))
+    unlist(bplapply(split(x, f), FUN = .estimate_precursor_intensity, ppm = ppm,
+                    tolerance = tolerance, method = method, msLevel = msLevel.,
+                    BPPARAM = BPPARAM), use.names = FALSE)
+}
+
+#' estimate precursor intensities based on MS1 peak intensity. This function
+#' assumes that `x` is a `Spectra` with data **from a single file/sample**.
+#'
+#' @author Johannes Rainer
+#'
+#' @importFrom stats approx
+#'
+#' @noRd
+.estimate_precursor_intensity <- function(x, ppm = 10, tolerance = 0,
+                                          method = c("previous",
+                                                     "interpolation"),
+                                          msLevel = 2L) {
+    method <- match.arg(method)
+    if (any(msLevel > 2))
+        warning("Estimation of precursor intensities for MS levels > 2 is ",
+                "not yet validated")
+    if (!length(x))
+        return(numeric())
+    if (is.unsorted(rtime(x))) {
+        idx <- order(rtime(x))
+        x <- x[idx]
+    } else idx <- integer()
+    pmz <- precursorMz(x)
+    pmi <- rep(NA_real_, length(pmz))
```

Comment:
It's a little counter intuitive, but:

```r
> library(microbenchmark)
> microbenchmark(rep(NA_real_, 1000), rep_len(NA_real_, 1000))
Unit: microseconds
                    expr min  lq  mean median   uq  max neval cld
     rep(NA_real_, 1000) 3.7 4.0 4.588    4.2 4.40 23.1   100  a 
 rep_len(NA_real_, 1000) 4.9 5.2 6.234    5.4 5.75 45.6   100   b
```

That's the reason I usually use `rep` instead of `rep_len`.

Reply:
Thanks, for the benchmark. Even `rep.int` is slower than `rep`, really counter intuitive.